uast/transformer/ops.go

```diff
@@ -0,0 +1,1099 @@
+func AnyVal(val uast.Value) Op {
+	return AnyNode(Is(val))
+}
+
+// And checks current node with all ops and fails if any of them fails.
+// Reversal applies all modifications from ops to the current node.
+// Typed ops should be at the beginning of the list to make sure that `Construct`
+// creates a correct node type before applying specific changes to it.
+func And(ops ...Op) Op {
+	if len(ops) == 1 {
+		return ops[0]
+	}
+	return opAnd(ops)
+}
+
+type opAnd []Op
+
+func (op opAnd) Check(st *State, n uast.Node) (bool, error) {
+	for i, sub := range op {
+		if ok, err := sub.Check(st, n); err != nil {
+			return false, errAnd.Wrap(err, i, sub)
+		} else if !ok {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (op opAnd) Construct(st *State, n uast.Node) (uast.Node, error) {
+	for i, sub := range op {
+		var err error
+		n, err = sub.Construct(st, n)
+		if err != nil {
+			return nil, errAnd.Wrap(err, i, sub)
+		}
+	}
+	return n, nil
+}
+
+var _ ObjectOp = Obj{}
+
+// Obj is a helper for defining a transformation on an object fields. See Object.
+// Operations will be sorted by the field name before execution.
+type Obj map[string]Op
+
+func (o Obj) Object() Object {
+	obj := Object{
+		set:    make(map[string]struct{}, len(o)),
+		fields: make(Fields, 0, len(o)),
+	}
+	for k, op := range o {
+		obj.set[k] = struct{}{}
+		obj.fields = append(obj.fields, Field{Name: k, Op: op})
+	}
+	sort.Sort(ByFieldName(obj.fields))
+	return obj
+}
+func (o Obj) Check(st *State, n uast.Node) (bool, error) {
+	return o.Object().Check(st, n)
+}
+
+func (o Obj) Construct(st *State, n uast.Node) (uast.Node, error) {
+	return o.Object().Construct(st, n)
+}
+
+// ObjectOp is an operation that is executed on an object. See Object.
+type ObjectOp interface {
+	Op
+	Object() Object
+}
+
+// Part defines a partial transformation of an object.
+// All unused fields will be stored into variable with a specified name.
+func Part(vr string, o ObjectOp) ObjectOp {
+	obj := o.Object()
+	obj.other = vr
+	return obj
+}
+
+// Pre will execute provided field operation before executing the rest of operations for an object.
+func Pre(fields Fields, o ObjectOp) ObjectOp {
+	obj := o.Object()
+	if err := obj.setFields(fields...); err != nil {
+		panic(err)
+	}
+	arr := make([]Field, len(fields)+len(obj.fields))
+
+	i := copy(arr, fields)
+	copy(arr[i:], obj.fields)
+
+	obj.fields = arr
+	return obj
+}
+
+// Post will execute provided field operation after executing the rest of operations for an object.
+func Post(o ObjectOp, fields Fields) ObjectOp {
+	obj := o.Object()
+	if err := obj.setFields(fields...); err != nil {
+		panic(err)
+	}
+	arr := make([]Field, len(fields)+len(obj.fields))
+
+	i := copy(arr, obj.fields)
+	copy(arr[i:], fields)
+
+	obj.fields = arr
+	return obj
+}
+
+var _ ObjectOp = Fields{}
+
+// Fields is a helper for multiple operations on object fields with a specific execution order. See Object.
+type Fields []Field
```

Comment:
I would like an expanded explanation about when to use Obj and when to use Fields in the annotations.